ext/drreg/drreg.c

```diff
@@ -173,15 +173,38 @@ get_where_app_pc(instr_t *where)
  * SPILLING AND RESTORING
  */
 
+/* Returns whether the given ilist has an existing usage of the given slot
+ * on or after `where`. Such a usage, if any, must have been added by a
+ * previous instrumentation pass, and tells us that it is not safe to reuse
+ * the slot in the current pass.
+ */
+static bool
+has_pending_slot_usage_by_prior_pass(per_thread_t *pt, instrlist_t *ilist, instr_t *where,
+                                     uint slot)
+{
+    if (!TEST(DRREG_HANDLE_MULTI_PHASE_SLOT_RESERVATIONS, pt->bb_props))
+        return false;
+    for (instr_t *in = where; in != NULL /*sanity*/ && in != instrlist_last(ilist);
```

Comment:
Why exclude last?  Why not just check for null?  It seems like a drreg restore could be the last instr?

Reply:
Thanks for catching this. I assumed usage like the [C++ std containers](https://en.cppreference.com/w/cpp/container/vector/end) where you stop iterating before the last element. But I see now in documentation of [`instrlist_last`](https://github.com/DynamoRIO/dynamorio/blob/1640ffe36d46ad68af5f1df67898d1ae069b056f/core/ir/instrlist_api.h#L143) that it returns the actual last element.

This usage also looks suspicious and may warrant auditing: https://github.com/DynamoRIO/dynamorio/blob/1640ffe36d46ad68af5f1df67898d1ae069b056f/core/vmareas.c#L4750

Reply:
That vmareas loop looks like it wants to stop when it hits the final jmp without the loop body looking at it (it's examined afterward), so the !=last seems correct.

Reply:
Okay. Thanks for checking. 

Done.